File: rubySrc/knapsack.rb
def knapsack(myKnapsack)
  Integer i = 0
  Integer itemNum = myKnapsack.getItemNum()
  Array items = Array.new(itemNum){ Array.new(2) }
  while i < itemNum
    Array item = myKnapsack.getItem(i)
    items[i][0] = item[0]
    items[i][1] = item[1]
    i = i + 1
  end
  Integer capacity = myKnapsack.getCapacity()
  Array dp = Array.new(itemNum + 1){ Array.new(capacity + 1) }
  i = 0
  while i < capacity + 1
    dp[0][i] = 0
    i = i + 1
  end
  i = 0
  Integer j = 0
  #p items
  while i < itemNum
    j = 0
    while j < capacity + 1
      if items[i][0] <= j then
        if items[i][1] + dp[i][j - items[i][0]] > dp[i][j] then
          dp[i + 1][j] = items[i][1] + dp[i][j - items[i][0]]
        else
          dp[i + 1][j] = dp[i][j]
        end
      else
        dp[i + 1][j] = dp[i][j]
      end
      j = j + 1
    end
    i = i + 1
  end

  i = itemNum
  j = capacity
  while i > 0
    if dp[i][j] != dp[i - 1][j] then
      myKnapsack.pickItem(i - 1)
      #p i-1
      #p items[i - 1]
      while dp[i][j] == dp[i][j - 1]
        j = j - 1
      end
      j = j + 1 - items[i - 1][0]
    end
    i = i - 1
  end
end

File: src/Knapsack.java


public class Knapsack {
	private int[][] item = {{168, 496}, {10, 45}, {145, 325}, {60, 347}, {10, 61}, {124, 486}, {124, 446}, {105, 22}, {126, 110}, {184, 475}};
	private int[][] returnItem;	//itemがRuby側で書き換えられないようにするため
	//1,4,5,6?
	private int capacity = 300;

	private boolean[] selectItem;
	private int myCapacity = 0;

	private boolean failFlag = false;		//失敗したとき

	public Knapsack() {
		selectItem = new boolean[item.length];
		for(int i = 0; i < selectItem.length; i++){
			selectItem[i] = false;
		}
		returnItem = new int[item.length][item[0].length];
	}

	public int getCapacity(){
		return capacity;
	}

	public int[] getItem(int i){
		returnItem[i][0] = item[i][0];
		returnItem[i][1] = item[i][1];
		return returnItem[i];
	}

	public int getItemNum(){
		return item.length;
	}

	public void pickItem(int i){
		if(failFlag){
			return;
		}
		if(myCapacity > capacity){	//失敗
			failFlag = true;
			return;
		}
		if(selectItem[i]){	//失敗
			failFlag = true;
			return;
		}

		selectItem[i] = true;
	}

	public void printResult(){
		System.out.println("正常終了:" + !failFlag);
		int c = 0, v = 0;
		for(int i = 0; i < selectItem.length; i++){
			if(selectItem[i]){
				System.out.println("選択:" + i + " 価値:" + item[i][1] + " 重さ:" + item[i][0]);
				c = c + item[i][0];
				v = v + item[i][1];
			}
		}
		System.out.println("価値合計:" + v);
		System.out.println("総重量:" + c);
	}
}
